app/waf/src/controllers/adc.controller.ts

```diff
@@ -751,3 +715,131 @@ type AddonReqValues = {
   userToken: string;
   tenantId: string;
 };
+
+class AdcStateCtrlr {
+  private states: AdcStateEntry[] = [
+    {
+      state: AdcState.NEW,
+      check: Promise.resolve(true),
+      next: [AdcState.POWERON],
+    },
+    {
+      state: AdcState.POWERON,
+      check: this.accessible(),
+      next: [AdcState.ONBOARDED, AdcState.RECLAIMED],
+    },
+    {
+      state: AdcState.ONBOARDED,
+      check: this.onboarded(),
+      next: [AdcState.TRUSTED, AdcState.RECLAIMED],
+    },
+    {
+      state: AdcState.TRUSTED,
+      check: this.trusted(),
+      next: [AdcState.RECLAIMED],
+    },
+    {
+      state: AdcState.RECLAIMED,
+      check: Promise.resolve(this.reclaimed()),
+      next: [AdcState.POWERON],
+    },
+  ];
+
+  constructor(private adc: Adc, private addon: AddonReqValues) {}
+
+  async readyTo(state: string): Promise<boolean> {
+    let stateEntry = this.getStateEntry(this.adc.status);
+    return (await stateEntry.check) && stateEntry.next.includes(state);
+  }
+
+  async gotTo(state: string): Promise<boolean> {
+    return this.getStateEntry(state).check;
+  }
+
+  private getStateEntry(name: string): AdcStateEntry {
+    return this.states.find(s => {
+      return s.state === name;
+    })!;
+  }
+
+  private reclaimed(): boolean {
+    if (this.adc.management) return false;
+    if (this.adc.compute.vmId) return false;
+    for (let net of Object.keys(this.adc.networks)) {
+      if (this.adc.networks[net].portId) return false;
+    }
+    return true;
+  }
+
+  private async trusted(): Promise<boolean> {
+    return Promise.resolve(true);
```

Comment:
don't we need to add logic to check if it is trusted or not here?

Reply:
We need to add this part. Because they are in 2 separate PRs, we may need to implement it later.